Add mediation support

chimp1984 opened the pull request.

chimp1984 committed:
Refactoring: Move arbitration package inside dispute package

chimp1984 committed:
Use abstract base class DisputeResolver for arbitrator

chimp1984 committed:
Refactoring: Move mediator to mediator package.

chimp1984 committed:
Let Mediator inherit DisputeResolver.

chimp1984 committed:
Do not use protobuf inheritance

- Do not use protobuf inheritance for Arbitrator and Mediator as it
would break backward compatibility (and protobuf inheritance sucks
anyway)

chimp1984 committed:
Refactoring: Move ArbitratorModule to parent package

chimp1984 committed:
Refactoring: Rename ArbitratorModule to DisputeModule

chimp1984 committed:
Add mediators to Filter

chimp1984 committed:
Add mediators to filter window

chimp1984 committed:
Use abstract DisputeResolverService as base class for ArbitratorService

- Add common base class for ArbitratorService and MediatorService

chimp1984 committed:
Fix test

chimp1984 committed:
Use abstract DisputeResolverManager as base class for ArbitratorManager

- Add common base class for ArbitratorManager and MediatorManager

chimp1984 committed:
Refactor: Move arbitratorregistration package inside register pkg

chimp1984 committed:
Refactor: Rename arbitratorregistration package to arbitrator

chimp1984 committed:
Add registration view for mediator

- With cmd+D one can open the mediator regisration in account screen.
For arbitrator its cmd+R

chimp1984 committed:
Separate pub key list for mediator (no new keys added yet)

chimp1984 committed:
Set new pubkeys for mediator registration

- Before release set new keys from maintainer who manages keys

chimp1984 committed:
Set disputes @Nullable. Add null checks

chimp1984 committed:
Remove pre v0.9 handling for supported arbitrators from offer

- We changed handling of arbitrator selection with v0.9 so the
supported arbitrators in the offer is not used anymore. As we
enforced v1.2 a while back for trading we can be sure no pre v0.9
clients are used anymore and we can remove the optional code part.

chimp1984 committed:
Remove supported arbitrators info in offer details window

- As we do not use supported arbitraors in offer anymore since v0.9
we can remove that.

chimp1984 committed:
Remove check for matching arbitrator languages

As we do not use the supported arbitratos from offer since v0.9 we can
remove that check.

chimp1984 committed:
Remove not used classes

chimp1984 committed:
Remove checks for arbitrator and mediator in offer

We do not use those fields anymore. We still need to keep the fields
not nullable as old clients have the check still.

chimp1984 committed:
Add check if sig of proto object is not empty

We got in dev testing sometimes an empty protobuf Alert. Might be
caused from protobuf copatibility issues during development but not
100% clear.
As it causes an exception and corrupted user db file we prefer to set
it to null.

chimp1984 committed:
Remove TakerSelectMediator

This is not used anymore. Currently we would get an exception in the
trade but with follow up changes we will fix that...

Mediator handling and selection will be done the same way like
arbitrator. The current mediator handling was a relict from earlier
partial support for mediators which never got completed. As still a
null check is in place we need to ensure backward compatibility.

chimp1984 committed:
Set arbitratorNodeAddresses and mediatorNodeAddresses to deprecated

We do not use arbitratorNodeAddresses and mediatorNodeAddresses anymore
but as there is a null check we still need to keep the field ans set it
to an empty arrayList.

chimp1984 committed:
Make ArbitratorSelection generic. Add MEDIATOR_ADDRESS

We want to use the same selection algorithm for mediators as for
arbitrators, so we make ArbitratorSelection generic.
We add MEDIATOR_ADDRESS as extraMap entry to TradeStatistics2 to be
able to track number of trades with specific mediators.

ExtraMap is used to add new data to existing protobuf definitions which
is supported also by not updated clients. Adding a new protobuf field
would only be supported by new clients. As mediator support is a new
feature we could add a new field but to keep it in the same style like
arbitrator we prefer to use the map here as well.

chimp1984 committed:
Refactor: Rename ArbitratorSelection to DisputeResolverSelection

chimp1984 committed:
Add mediator to OfferAvailabilityResponse and mediatorNodeAddress to OpenOffer

WIP for supporting mediator selection the same way like arbitrators.

chimp1984 committed:
Make arbitrator not nullable

We can ensure that all users are post v0.9 so we can remove the nullable
 support.

chimp1984 committed:
Add selectedMediator to OfferAvailabilityModel

Remove nullable support in ProcessOfferAvailabilityResponse as we can
ensure all clients are post v0.9

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Add todo for using more generic keys for display strings

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Fix wrong handling of registeredMediator

Fix copy/paste error

chimp1984 committed:
Add mediatorNodeAddress to trade

chimp1984 committed:
Handle nullable mediator in ProcessOfferAvailabilityResponse

We do not get the mediator set from old clients but we expect a not null
value so we use the DisputeResolverSelection in case it is null.
We need to pass mediatorManager and tradeStatisticsManager to the
OfferAvailabilityModel.

chimp1984 committed:
Change log level, cleanup

chimp1984 committed:
Revert changes in OfferPayload due backward compatibility issues

Because of backward compatibility issues we needed to revert the removal
of arbitratorNodeAddresses and mediatorNodeAddresses. The signature
check for the offer  would fail as an old client would send a not-empty
list but new clients would have had an empty list, so the hash
would be different and the sig check fail and we would not accept that
offer. That is the reason why we still need to support those data even
it is not used anymore.

This is one of the more tricky cases for backward compatibility issues.

This version now is tested between new and old clients and trade and
disputes work.

chimp1984 committed:
Add checks if any mediator is available

chimp1984 committed:
Cleanup classes

chimp1984 committed:
Fix test

chimp1984 committed:
Add mediator DisputeStates

Add isMediationDispute to Dispute class.

If a dispute opening gets requested we check if state is
DisputeState.NO_DISPUTE and the open mediation. If state is
DisputeState.MEDIATION_REQUESTED we open arbitration.

chimp1984 committed:
Cleanup; support isMediationDispute

chimp1984 committed:
Handle mediator data in Dispute domain

- Add getConflictResolverNodeAddress method to Dispute to resolve
arbitrator or mediator address based on isMediationDispute flag.

- Rename arbitratorPubKeyRing to conflictResolverPubKeyRing in Dispute.
We cannot rename arbitratorPubKeyRing in the protobuf definition
as it would break backward compatibility.

chimp1984 committed:
Add support for mediation in dispute domain

- Add isMediationDispute method to ChatSession
- Add isMediationDispute method to DisputeCommunicationMessage
- Add isMediationDispute to dispute id
- Refactor findDispute method
- Add null checks
- Cleanups

chimp1984 committed:
Remove impossible case

Reserved and locked funds are used for offers and trades only.

chimp1984 committed:
Fix typos

chimp1984 committed:
Handle mediator and arbitrator strings

- Work in progress of adjusting correct terms.
- Cleanups

chimp1984 committed:
Refactor: Rename arbitrator package to disputeresolvers

chimp1984 committed:
Refactor: Rename ArbitratorDisputeView classes to DisputeResolverView

chimp1984 committed:
Add support for close ticket from mediator (WIP)

In mediator case we do not create any transaction but only send the
dispute result which contains the mediators recommended payout
distribution. At teh traders we set the disputeState in the trade to
closed. This will be used in the next commits to update the trade so
that the traders get displayed the recommended payout and get asked if
they agree to that.

chimp1984 committed:
Refactoring: Rename class

Rename MessageDeliveryFailedException to
DisputeMessageDeliveryFailedException

chimp1984 committed:
Refactoring: Move dispute classes to dispute package

chimp1984 committed:
Refactoring: Move Attachment class to dispute package

chimp1984 committed:
Refactoring: Move package one level up

Move bisq.core.dispute.arbitration.messages to
bisq.core.dispute.messages

chimp1984 committed:
Add todo comment

chimp1984 committed:
Use ARBITRATION instead of DISPUTE

chimp1984 committed:
Make DisputeManager abstract base class for ArbitrationDisputeManager

WIP for separating DisputeManager to ArbitrationDisputeManager and
MediationDisputeManager

chimp1984 committed:
Add MediationDisputeManager

chimp1984 committed:
Add MediationDisputeManager and ArbitrationDisputeManager to test

chimp1984 committed:
Add mediationDisputeManager to relevant classes

There are some cases where arbitrationDisputeManager only is used.
Those are usually related to the payout tx. As mediators do not do a
payout we don't need it there.

chimp1984 committed:
Add TradersArbitrationDisputeView and TradersMediationDisputeView

WIP for separating TraderDisputeView

chimp1984 committed:
Refactor:  Rename class

chimp1984 committed:
Refactor: Rename support.tab.support to support.tab.mediation.support

I am aware that committing non default translation files is not
recommended, but I think in that case it helps to avoid to show errors
for developers who use non-english locale. The changes will be
overwritten by transifex once it gets synced...

chimp1984 committed:
Add DisputeView as common base class

Further refactor separation of diff. dispute views

chimp1984 committed:
Refactor: Rename package

chimp1984 committed:
Refactor: Rename DisputesView to SupportView

chimp1984 committed:
Refactor: Rename package

chimp1984 committed:
Add MediationDisputeManager to CorePersistedDataHost

chimp1984 committed:
Add MediationDisputeList as db file, refactor DisputeList

WIP for making Dispute domain more generic. We want to separate
arbitration and mediation clearly.

chimp1984 committed:
Further refactoring to split mediation and arbitration

chimp1984 committed:
Further refactoring to split mediation and arbitration

Move methods used for arbitration only to ArbitrationDisputeManager

chimp1984 committed:
Refactor: Rename package

Rename bisq.core.dispute to bisq.core.support

No other changes in that commit.
We want to improve the data structure with the trader chat.
Support will be the top level.
Then dispute containing arbitration and mediation.
Next to dispute will be trader chat.

bisq.core.support
bisq.core.support.dispute.arbitration
bisq.core.support.dispute.mediation
bisq.core.support.traderchat  (not happy with name for that yet)

chimp1984 committed:
Refactor: Move dispute domain classes into isq.core.support.dispute package

chimp1984 committed:
Refactor: Move classes

Move bisq.core.chat.ChatSession to bisq.core.support.ChatSession
Move bisq.core.chat.ChatManager to bisq.core.support.ChatManager
Move bisq.core.trade.TradeChatSession to bisq.core.support.traderchat.TradeChatSession

chimp1984 committed:
Refactor: Move DisputeCommunicationMessage

chimp1984 committed:
Refactor: Rename DisputeCommunicationMessage to ChatMessage

chimp1984 committed:
Add comments

chimp1984 committed:
Refactor: Move class

chimp1984 committed:
Refactor: Rename class

chimp1984 committed:
Refactor: Rename addDisputeCommunicationMessage and strings and variables

Rename disputeCommunicationMessage to chatMessage

chimp1984 committed:
Refactor: Rename method

chimp1984 committed:
Refactor: Rename methods and strings

chimp1984 committed:
Add ArbitrationChatMessage and DisputeChatMessage

chimp1984 committed:
Refactor: Rename class

chimp1984 committed:
Move ChatMessage.Type to SupportType

Add to all supportMessages the SupportType so that we can filter in our
chatSessions the messages we are interested in.

chimp1984 committed:
Refactor: Move classed to new package

chimp1984 committed:
Refactor: Rename package

chimp1984 committed:
Refactor: Move classed to new package

chimp1984 committed:
Refactor: Move classed to new package

chimp1984 committed:
Refactor: Rename classes

chimp1984 committed:
Refactor: Rename package

chimp1984 committed:
Refactor: Rename classes

chimp1984 committed:
Refactor: Rename classes

chimp1984 committed:
Remove empty DisputeModule

chimp1984 committed:
Refactor: Rename classes

chimp1984 committed:
Refactor SupportManager domain (WIP)

chimp1984 committed:
Refactor SupportSession domain (WIP)

chimp1984 committed:
Remove methods from SupportSession

chimp1984 committed:
Dont expose p2pService in SupportManager

chimp1984 committed:
Remove supportType in SupportSession

chimp1984 committed:
Remove supportSession from getPeerNodeAddress method

chimp1984 committed:
Remove isBuyer from supportSession

chimp1984 committed:
Move creation of ChatMessage to SupportManager

chimp1984 committed:
Remove isMediationDispute fielf in ChatMessage

chimp1984 committed:
Remove chatMessage.isMediationDispute()

chimp1984 committed:
Refactor: Rename trade.getCommunicationMessages()

chimp1984 committed:
Move creation of ChatMessage to Chat

chimp1984 committed:
Refactor: Rename class

chimp1984 committed:
Refactor: Move ChatView class

chimp1984 committed:
Refactor: Move PriceFeedComboBoxItem class to shared package

chimp1984 committed:
Refactor: Use 'public abstract' instead of 'abstract public'

chimp1984 committed:
Refactor: Use 'protected abstract' instead of 'abstract protected'

chimp1984 committed:
Add traderChatManager.onAllServicesInitialized() to BisqSetup

chimp1984 committed:
Remove unused param

chimp1984 committed:
Refactor: Rename addChatMessage to addAndPersistChatMessage

chimp1984 committed:
Fix missing check at ack msg handling

Various WIP refactorings/improvements

chimp1984 committed:
Remove addAndPersistChatMessage from SupportSession

chimp1984 committed:
Remove disputeManager from DisputeSession

chimp1984 committed:
Fix missing getConcreteDisputeChatSession impl.

chimp1984 committed:
Refactor: Rename package

chimp1984 committed:
Refactor: Rename classes

Avoid trader as it might confuse with trader chat.
As for mediation/arbitration the agent (mediator/arbitrator) are acting
a bit like a server we use the client terminology for the traders.

chimp1984 committed:
Refactor: Move classes to new package

chimp1984 committed:
Fix missing protobuf data

- Add missing SupportType to protobuf
- Remove is_mediation_dispute from Dispute protobuf
 definition
- Add getAgentNodeAddress method
- Var. other refactorings, cleanups

chimp1984 committed:
Clone list at persisting to avoid ConcurrentModificationException

chimp1984 committed:
Fix order of SupportType

Old clients fall back to enum at slot 0.

chimp1984 committed:
Add getDisputeState_StartedByPeer template method

chimp1984 committed:
Add trade protocol tasks for mediation result tx signing and msg sending

chimp1984 committed:
Complete protocol tasks for mediation

chimp1984 committed:
Refactor: Remove unneeded SuppressWarnings type: "WeakerAccess"

chimp1984 committed:
Complete mediation result protocol

Works now all but not much tested....

chimp1984 committed:
Add activation date and capability

We need to make sure that not updated users cannot cause problems once mediation is supported. We would get mixed cases where one has a mediation ticket and the not updated user an arbitration ticket. To avoid that we set an activation date with about 10 days from release. Until that date mediation is not supported.
Additionally we use OfferRestrictions.REQUIRE_UPDATE_DATE for hiding offers from users how have not updated (we use the fact that mediator and arbitrator has been same in old version, in new version they are different).
An old client cannot take an offer from a new maker as he does not has set the new MEDIATION capability. He will get an null value as AvailabilityResult as he has not the new entry MISSING_MANDATORY_CAPABILITY.
We will also use the min version for trading in the filter, so that not updated users get a popup telling them to update and they see all offers deactivated.

chimp1984 committed:
Various fixes

chimp1984 committed:
Remove code part which does not make sense (anymore)

Maybe in older versions there was use of openDisputes and closedDisputes
but now it does not make sense anymore and arbitrator never gets 4 cases
opened if offline.

chimp1984 committed:
Add check of balance is > 0

chimp1984 committed:
Only close trade if payout tx is set

chimp1984 committed:
Add missing check if arbitrator and mediator are available

chimp1984 committed:
Fix wrong key

chimp1984 committed:
Improve handling of checks and popup display

For create and take offer we check certain conditions and show a
popup if not met. This commit moves that to GuiUtils.

chimp1984 committed:
Rename any occurrance of DisputeResolver to DisputeAgent

chimp1984 committed:
Fix handling of mediatorPubKeyRing

chimp1984 committed:
Remove disputeSummaryWindow.evidence fields

chimp1984 committed:
Add missing persistence for MediationResultState

chimp1984 committed:
Fix tests

chimp1984 committed:
Make text more compact to not exceed space

chimp1984 committed:
Refactor NotificationGroup

chimp1984 committed:
Improve text, add dev testing feature for popups

chimp1984 committed:
Improve text

chimp1984 committed:
Renamed a key and assigned a new text

chimp1984 committed:
Fix states

chimp1984 committed:
Do not set errorMessage

Do not set errorMessage if both peers have opened a dispute and agent
was not online

chimp1984 commented:
For testing trade period half over and trade period over popups and states I added a keyboard shortcut: cmd + up for going from first half -> second half -> trade period over and cmd + down for the other direction. It only shows state changes when not in mediation or dispute mode. But after a change and a restart a period warning popup is displayed.
This dev test feature is only active for v 1.1.5, but code should be removed before release.

I set the activation dates in OfferRestrictions and MediationManager to 1st of september, so mediation is fully activated. For release we need to set date for OfferRestrictions 3-5 days after release date (after that date not updated users cannot take an offer and their offers will be filtered by updated nodes). We will set additionally a filter to enforce update. That shows a popup for updating as well.
Activation date in MediationManager is the date when the mediation will be enabled. Before that opening a dispute will open arbitration. We should set taht data about 1 week after OfferRestrictions.REQUIRE_UPDATE_DATE so that pending trades get completed and to avoid mixed cases where one peer opens mediation with new version but old opens arbitration with old version.

This PR is a huge code change and all the tradeprotocol, mediation, arbitration and trade chat need to be tested very carefully. There are many possible variations. I will test over the weekend further so monday it should be battletested from my side.

For reviewing the PR I recommend to read the full tradeprotocol, mediation, arbitration and trade chat domain. Too many chnages that it is feasible to go from commit to commit...

The shortcut for registering an arbitrator is cmd + r at the account screen and for mediator cmd +d (only works with `--useDevPrivilegeKeys=true`). You cannot register both in one app. So you need 2 instances. 

Do NOT use `--devMode=true` as it avoids lots of popups and notifications so testing would miss a lot.

chimp1984 commented:
<img width="2396" alt="Screen Shot 2019-09-06 at 23 44 34" src="https://user-images.githubusercontent.com/54558767/64462520-ff9a0580-d100-11e9-9cf8-8ba739cbaff6.png">
<img width="1189" alt="Screen Shot 2019-09-06 at 23 45 13" src="https://user-images.githubusercontent.com/54558767/64462522-ff9a0580-d100-11e9-8773-e97bb4a1e92b.png">
<img width="2397" alt="Screen Shot 2019-09-06 at 23 45 51" src="https://user-images.githubusercontent.com/54558767/64462523-00329c00-d101-11e9-8a2c-dab77a5f736b.png">
<img width="1200" alt="Screen Shot 2019-09-06 at 23 46 05" src="https://user-images.githubusercontent.com/54558767/64462524-00329c00-d101-11e9-9dca-e46739fe9675.png">
<img width="2401" alt="Screen Shot 2019-09-06 at 23 46 59" src="https://user-images.githubusercontent.com/54558767/64462525-00329c00-d101-11e9-8b51-138196d32186.png">
<img width="1201" alt="Screen Shot 2019-09-06 at 23 47 10" src="https://user-images.githubusercontent.com/54558767/64462526-00329c00-d101-11e9-9c27-b9587be28a15.png">
<img width="2398" alt="Screen Shot 2019-09-06 at 23 47 46" src="https://user-images.githubusercontent.com/54558767/64462527-00329c00-d101-11e9-99ba-958a0dd5a7c7.png">
<img width="1198" alt="Screen Shot 2019-09-06 at 23 48 37" src="https://user-images.githubusercontent.com/54558767/64462528-00cb3280-d101-11e9-9a7c-be8be1b578f8.png">

chimp1984 committed:
Remove logs used for dev testing

christophsturm commented on common/src/main/proto/pb.proto:
```diff
@@ -44,6 +44,7 @@ message NetworkEnvelope {
 
         OpenNewDisputeMessage open_new_dispute_message = 22;
         PeerOpenedDisputeMessage peer_opened_dispute_message = 23;
+        // ChatMessage in java class. We cannot rename protobuf definition because it would break backward compatibility
```
does it really break compatibility? i think protobuf uses just the id.

chimp1984 replied:
I just tested it. With pure network messages the class and field name are irrelevant, so yes in that case we can rename it. I will make a commit later.
Though we have to take care to not do that on protobuf objects which are peristed (or used in both network and peristence context) as locally stored data would break then.

christophsturm replied:
because we use the class name in our local storage code?

chimp1984 replied:
Ups, was wrong thought... Yes is same with peristed data.

chimp1984 replied:
@christophsturm Enums must not be renamed as we use the name for resolving.

chimp1984 committed:
Fix getMedian method with empty list

chimp1984 committed:
Add new methods and tests

Add fromCommaSeparatedOrdinals and toCommaSeparatedOrdinals to convert
from string representations (used for handling backward compatibility
with mediation release).

Add check if int >= 0 to fromIntList

chimp1984 committed:
Move error log outside of delayed call

chimp1984 committed:
Add capabilities entry to extraDataMap in offer

The previous implementation did not work for supporting updates and
hiding offers from not updated clients.
We use now the capabilities converted to a string list and put it into
the extraDataMap. If a use with old persisted offers updates his offers
gets converted to add the capabilities. Updated clients will ignore
offers without the mediation capability set in the offer.

chimp1984 committed:
Rename non sync protobuf definitions

As Christoph Sturm pointed out we can rename protobuf entries.
Only index number must not be changed.

chimp1984 committed:
Fix UI state when arbitration has started

Only set mediation state if we are not in arbitration state.

chimp1984 commented:
Btw. I did not refactor much of the old code in the domain I touched to limit efforts and risks. So there are still very long and complex methods in the code which should be refactored some day...

chimp1984 committed:
Remove restriction

chimp1984 commented:
I tested quite a lot and did not find any issues anymore. So from my side it is ready for testing.

If anyone wants to help testing get in touch as there is a bit of background info required for the test setup...

chimp1984 committed:
Fix typo; remove errorMessage

If both have opened a dispute and agent was not online we dont treat it
as error.

chimp1984 committed:
Improve text

chimp1984 committed:
Store full address for localhost dev testing

The arbitrator/mediator selection is based on statistics of usage of
agents in past trades. We put the first 4 chars into the trade
statistics, but for localhost that would be same vale for 2 diff nodes.

chimp1984 committed:
Remove errorMessage

If both have opened a dispute and agent was not online we dont treat it
as error.

chimp1984 committed:
Improve text

chimp1984 commented:
Here is a known issues with a scenario in the update process which is not handled.

Case: 
- Alice has old version and Bob has updated and mediation is already activated.
- Selected arbitrator and mediator in trade are different (2 arbitrators, and mediators registered)
- Bob opens dispute
- Alice will get arbitration as she is on old version. Bob gets mediation. Once mediator closes it looks for Alice like arbitrator has closed and trade gets closed, but there is no payout. Bob can accept mediation but Alice will never respond, so at some point Bob will open an arbitration ticket and the the arbitrator will do the payout and Alice will get the payout tx as well.

This scenario should be very rare and if users update quickly it will hopefully not happen. As at least one party is aware the the payout is pending it is also low risk the the payout will never be made. Also Alice should check if she really got the payout when her trade was closed. Also the mediator will see the version of Alice and can tell her to update as well that this case might not behave as expected.

chimp1984 committed:
Keep accept or reject button enabled after accept

- If peer never accepts the trader who has accepted first can change
to reject to open a arbitration dispute.
We could improve that by adding a new state to open arbitration
directly and show a diff. button text and popup. But I think for now
thats ok as well....

chimp1984 committed:
Cleanups (no functional change)

- remove unused params
- remove not used code
- reformat
- clean up comments
- fix log levels
- remove redundant annotations

ripcurlx commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
+
+portfolio.pending.mediationResult.headline=Suggested payout from mediation
+portfolio.pending.mediationResult.info.noneAccepted=Complete the trade by accepting the mediator's suggestion for the trade payout.
+portfolio.pending.mediationResult.info.selfAccepted=You have accepting the mediator's suggestion. Waiting for peer to accept as well.
```
@chimp1984 Great work as always! 👍 Before I squash and merge the commits (just too many commits to overlook everything) I'd like to have a text review by @m52go before. @m52go Could you please go over the new translations for the mediation system? e.g. here I think it should be `You have accepted the mediator's sugestion....`

m52go replied:
```suggestion
portfolio.pending.mediationResult.info.selfAccepted=You have accepted the mediator's suggestion. Waiting for peer to accept as well.
```

ripcurlx commented:
> ![](https://user-images.githubusercontent.com/54558767/64462526-00329c00-d101-11e9-9c27-b9587be28a15.png)

Here I think the wording needs to be improved. The text is asking for a confirmation and in the button it says `Accept or Reject`. Feels a little bit as I'm at a roulette table as it is not 100% clear if this button will trigger an undesired action. As we need to have the detailed mediation result acceptance in that case I would label it something like `Review Payout Suggestion`. What do you think?

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -541,8 +544,9 @@ portfolio.pending.step3_seller.confirmPaymentReceived=Confirm payment received
 portfolio.pending.step1.info=Deposit transaction has been published.\n{0} need to wait for at least one blockchain confirmation before starting the payment.
-portfolio.pending.step1.warn=The deposit transaction still did not get confirmed.\nThat might happen in rare cases when the funding fee of one trader from the external wallet was too low.
-portfolio.pending.step1.openForDispute=The deposit transaction still did not get confirmed.\nThat might happen in rare cases when the funding fee of one trader from the external wallet was too low.\nThe max. period for the trade has elapsed.\n\nYou can wait longer or contact the arbitrator for opening a dispute.
+portfolio.pending.step1.warn=The deposit transaction is still not confirmed. That might happen in rare cases when the funding fee of one trader from the external wallet was too low.
```
```suggestion
portfolio.pending.step1.warn=The deposit transaction is still not confirmed. This sometimes happens in rare cases when the funding fee of one trader from an external wallet was too low.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -605,21 +613,21 @@ portfolio.pending.step2_buyer.confirmStart.yes=Yes, I have started the payment
 portfolio.pending.step2_seller.waitPayment.msg=The deposit transaction has at least one blockchain confirmation.\nYou need to wait until the BTC buyer starts the {0} payment.
-portfolio.pending.step2_seller.warn=The BTC buyer still has not done the {0} payment.\nYou need to wait until they have started the payment.\nIf the trade has not been completed on {1} the arbitrator will investigate.
-portfolio.pending.step2_seller.openForDispute=The BTC buyer has not started his payment!\nThe max. allowed period for the trade has elapsed.\nYou can wait longer and give the trading peer more time or contact the arbitrator for opening a dispute.
+portfolio.pending.step2_seller.warn=The BTC buyer still has not done the {0} payment.\nYou need to wait until they have started the payment.\nThe trade has to be completed until {1}.
```
Is {1} the date by which the trade needs to be completed by? If so, I suggest this.

```suggestion
portfolio.pending.step2_seller.warn=The BTC buyer still has not done the {0} payment.\nYou need to wait until they have started the payment.\nThe trade has to be completed by {1}.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -605,21 +613,21 @@ portfolio.pending.step2_buyer.confirmStart.yes=Yes, I have started the payment
 tradeChat.rules=You can communicate with your trade peer to resolve potential problems with this trade.\n\
   It is not mandatory to reply in the chat.\n\
-  If a trader violates the below rules, open a dispute with 'Cmd/Ctrl + o' and report it to the arbitrator.\n\n\
+  If a trader violates the below rules, open a dispute and report it to the mediator or arbitrator.\n\n\
```
```suggestion
  If a trader violates any of the rules below, open a dispute and report it to the mediator or arbitrator.\n\n\
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -640,8 +648,10 @@ portfolio.pending.step3_buyer.wait.info=Waiting for the BTC seller''s confirmati
 portfolio.pending.step3_buyer.warn.part1b=at your payment provider (e.g. bank)
-portfolio.pending.step3_buyer.warn.part2=The BTC seller still has not confirmed your payment!\nPlease check {0} if the payment sending was successful.\nIf the BTC seller does not confirm the receipt of your payment by {1} the trade will be investigated by the arbitrator.
-portfolio.pending.step3_buyer.openForDispute=The BTC seller has not confirmed your payment!\nThe max. period for the trade has elapsed.\nYou can wait longer and give the trading peer more time or contact the arbitrator for opening a dispute.
+portfolio.pending.step3_buyer.warn.part2=The BTC seller still has not confirmed your payment yet. Please check {0} if the \
```
```suggestion
portfolio.pending.step3_buyer.warn.part2=The BTC seller still has not confirmed your payment. Please check {0} if the \
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
 portfolio.pending.tradeNotCompleted=Trade not completed in time (until {0})
 portfolio.pending.tradeProcess=Trade process
-portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the arbitrator arrived (e.g. if you did not get a response after 1 day) feel free to open a dispute again.
+portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the mediator or arbitrator arrived \
```
Should we mention the shortcut here (Cmd/Ctrl+o)?

Also should we mention the forum URL (https://bisq.community)?

Suggestion:

```
portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the mediator or arbitrator arrived / (e.g. if you did not get a response after 1 day) feel free to open a dispute again with Cmd/Ctrl+o. You can also ask for additional help on the Bisq forum at https://bisq.community.
```

chimp1984 replied:
Yes, agree. Can you make another suggestion so I can commit here?

chimp1984 replied:
Applied your text in next commit.

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
 portfolio.pending.openAgainDispute.button=Open dispute again
 portfolio.pending.openSupportTicket.headline=Open support ticket
-portfolio.pending.openSupportTicket.msg=Please use that only in emergency case if you don't get displayed a \"Open support\" or \"Open dispute\" button.\n\nWhen you open a support ticket the trade will be interrupted and handled by the arbitrator
+portfolio.pending.openSupportTicket.msg=Please use that only in emergency case if you don't get displayed a \
```
```
portfolio.pending.openSupportTicket.msg=Please use this function only in emergency cases if you don't see a \
\"Open support\" or \"Open dispute\" button.\n\nWhen you open a support ticket the trade will be interrupted and \
handled by a mediator or arbitrator.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
+portfolio.pending.support.headline.getHelp=Need help?
+portfolio.pending.support.text.getHelp=If you have any problems you can try to contact the trade peer in the trade \
+  chat or ask the Bisq community at https://bisq.community. \
+  If you cannot resolve your problems you can request help from the mediator.
```
```suggestion
  If your issue still isn't resolved, you can request more help from a mediator.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
+  chat or ask the Bisq community at https://bisq.community. \
+  If you cannot resolve your problems you can request help from the mediator.
+portfolio.pending.support.button.getHelp=Get support
+portfolio.pending.support.popup.info=If you cannot resolve your problems with the trade peer you can open a support \
```
```
portfolio.pending.support.popup.info=If your issue with the trade remains unsolved, you can open a support \ ticket to request help from the mediator. If you have not received the payment, please wait until the trade period is over.\n\n\ Are you sure you want to open a support ticket?
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
+portfolio.pending.mediationResult.headline=Suggested payout from mediation
+portfolio.pending.mediationResult.info.noneAccepted=Complete the trade by accepting the mediator's suggestion for the trade payout.
+portfolio.pending.mediationResult.info.selfAccepted=You have accepting the mediator's suggestion. Waiting for peer to accept as well.
+portfolio.pending.mediationResult.info.peerAccepted=Your trade peer has accepting the mediator's suggestion. Do you accept as well?
```
```suggestion
portfolio.pending.mediationResult.info.peerAccepted=Your trade peer has accepted the mediator's suggestion. Do you accept as well?
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
+portfolio.pending.mediationResult.info.peerAccepted=Your trade peer has accepting the mediator's suggestion. Do you accept as well?
+portfolio.pending.mediationResult.button=Accept or reject
+portfolio.pending.mediationResult.popup.headline=Mediation result for trade with ID: {0}
+portfolio.pending.mediationResult.popup.headline.peerAccepted=Your trade peer has accepting the mediator''s suggestion for trade {0}
```
```suggestion
portfolio.pending.mediationResult.popup.headline.peerAccepted=Your trade peer has accepted the mediator''s suggestion for trade {0}
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -879,20 +934,15 @@ support.buyerOfferer=BTC buyer/Maker
-bisq.network web page or post in the Bisq forum in the Support section.
 
+# TODO @m52go could you provide a good text here?
+support.backgroundInfo=Bisq is not a company, so it handles disputes differently.\n\n\
```
```
support.backgroundInfo=Bisq is not a company, so it handles disputes differently.\n\n\
Traders can communicate within the application via a secure chat on the pending trades screen to attempt solving a dispute on their own. \
  If that is not sufficient, a mediator can step in to help. The mediator will evaluate the situation and give a recommendation for the \
  payout of the trade funds. If both traders accept this suggestion, the payout transaction is completed and the trade is closed. \
  If one or both traders do not agree to the mediator's recommended payout, they can request arbitration.\
  The arbitrator has the third key of the deposit transaction and will make the payout based on their findings.
```

chimp1984 replied:
Applied your text in next commit.

m52go reviewed (commented):
I tried to make suggestions into suggested edits wherever possible, but there were a handful of multi-line suggestions where that wasn't really practical.

Would be great if someone could take a quick look at my documentation change for mediation too. It's short: https://github.com/bisq-network/bisq-docs/pull/160

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
 portfolio.pending.tradeProcess=Trade process
-portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the arbitrator arrived (e.g. if you did not get a response after 1 day) feel free to open a dispute again.
+portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the mediator or arbitrator arrived \
+  (e.g. if you did not get a response after 1 day) feel free to open a dispute again. You can also ask on the \
```
```suggestion
  (e.g. if you did not get a response after 1 day) feel free to open a dispute again with Cmd/Ctrl+o. You can also ask \
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -732,27 +745,69 @@ portfolio.pending.tradePeriodInfo=After the first blockchain confirmation, the t
-portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the arbitrator arrived (e.g. if you did not get a response after 1 day) feel free to open a dispute again.
+portfolio.pending.openAgainDispute.msg=If you are not sure that the message to the mediator or arbitrator arrived \
+  (e.g. if you did not get a response after 1 day) feel free to open a dispute again. You can also ask on the \
+  Bisq forum web page to get feedback from the mediator or arbitrator.
```
```suggestion
  for additional help on the Bisq forum at https://bisq.community.
```

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Improve text

chimp1984 committed:
Merge branch 'add-mediation-support' of https://github.com/chimp1984/bisq into add-mediation-support

 Conflicts:
	core/src/main/resources/i18n/displayStrings.properties

chimp1984 committed:
Merge branch 'v1.1.6' into add-mediation-support

# Conflicts:
#	common/src/main/java/bisq/common/app/Capability.java
#	core/src/main/java/bisq/core/setup/CoreNetworkCapabilities.java

chimp1984 committed:
Auto fill remaining amount in custom payout

If mediator or arbitrator are doing a custom payout, we auto-fill
counterpart field with remaining amount, so he does not need to
calculate.

chimp1984 commented:
@ripcurlx Text changes are applied.

ripcurlx commented:
![Bildschirmfoto 2019-09-11 um 10 05 19](https://user-images.githubusercontent.com/170962/64679420-fd76d480-d47b-11e9-9dba-99cb178c04ec.png)
@chimp1984 Not critical, but I would remove the `Need help...` in the forth step to prevent unnecessary mediation cases (there shouldn't be any issues at that point that require mediation)

ripcurlx commented:
Although it will only be for 10 days as a transitional period that the mediation system will be activated, I think it will cause too much confusion if we have the mediation text in the the "Need help..." and within the popup, but open an arbitration case afterwards. 
![Bildschirmfoto 2019-09-11 um 10 47 25](https://user-images.githubusercontent.com/170962/64682906-9c063400-d482-11e9-92fa-c21d666ebdf2.png)
<img width="1330" alt="Bildschirmfoto 2019-09-11 um 10 53 45" src="https://user-images.githubusercontent.com/170962/64682907-9c9eca80-d482-11e9-83c8-6ba2918c17c9.png">

ripcurlx commented:
I don't think we have to get that far to deactivate the mediation tab in the support section as well, but within the trade I think it creates too much confusion.
<img width="1374" alt="Bildschirmfoto 2019-09-11 um 11 07 46" src="https://user-images.githubusercontent.com/170962/64683940-8265ec00-d484-11e9-8af2-f57b4b88b3e3.png">

ripcurlx commented:
> ![Bildschirmfoto 2019-09-11 um 10 05 19](https://user-images.githubusercontent.com/170962/64679420-fd76d480-d47b-11e9-9dba-99cb178c04ec.png)
> @chimp1984 Not critical, but I would remove the `Need help...` in the forth step to prevent unnecessary mediation cases (there shouldn't be any issues at that point that require mediation)

Actually we need to remove it here, as within arbitration it looks to the arbitrator as if he/she is able to change the output when closing the case, but actually it doesn't have any effect as the multisig was already signed and broadcasted.

ripcurlx commented:
### Scenario
Buyer opens dispute with old client version and receives payout
--Given Alice as buyer (maker), Bob as seller (taker), and arbitrator are online
--And Alice is running an old client version (v1.1.5)
--And Bob is running the current client release version (1.1.6)
--And Alice has published a buy offer

I've set `private static final Date REQUIRE_UPDATE_DATE = Utilities.getUTCDate(2019, GregorianCalendar.SEPTEMBER, 12);` and `private static final Date MEDIATION_ACTIVATED_DATE = Utilities.getUTCDate(2019, GregorianCalendar.SEPTEMBER, 12);`. This should be the case when we release 1.1.6 and mediation is not activated and no update is required for trading yet.

When Bob tries to take the offer from Alice, Alice sees following error message
![Bildschirmfoto 2019-09-11 um 14 13 42](https://user-images.githubusercontent.com/170962/64697428-1e9dec00-d4a1-11e9-9060-c98798db1999.png)
```
java.lang.NullPointerException: mediator must not be null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:787)
	at bisq.core.trade.Trade.setMediatorNodeAddress(Trade.java:702)
	at bisq.core.trade.protocol.tasks.maker.MakerProcessPayDepositRequest.run(MakerProcessPayDepositRequest.java:80)
	at bisq.common.taskrunner.TaskRunner.next(TaskRunner.java:69)
	at bisq.common.taskrunner.TaskRunner.run(TaskRunner.java:60)
	at bisq.core.trade.protocol.BuyerAsMakerProtocol.handleTakeOfferRequest(BuyerAsMakerProtocol.java:143)
	at bisq.core.trade.BuyerAsMakerTrade.handleTakeOfferRequest(BuyerAsMakerTrade.java:104)
	at bisq.core.trade.TradeManager.handlePayDepositRequest(TradeManager.java:346)
	at bisq.core.trade.TradeManager.lambda$new$0(TradeManager.java:180)
	at bisq.network.p2p.P2PService.lambda$onMessage$8(P2PService.java:410)
	at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:804)
	at java.base/java.util.concurrent.CopyOnWriteArraySet.forEach(CopyOnWriteArraySet.java:425)
	at bisq.network.p2p.P2PService.onMessage(P2PService.java:409)
	at bisq.network.p2p.network.NetworkNode.lambda$onMessage$4(NetworkNode.java:344)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at bisq.network.p2p.network.NetworkNode.onMessage(NetworkNode.java:344)
	at bisq.network.p2p.network.Connection.lambda$onMessage$4(Connection.java:428)
	at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:804)
	at java.base/java.util.concurrent.CopyOnWriteArraySet.forEach(CopyOnWriteArraySet.java:425)
	at bisq.network.p2p.network.Connection.lambda$onMessage$5(Connection.java:428)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
```
I was surprised that we had some mediator code in our old version already 🤔 
During the time of creating the offer there was already a mediator and an arbitrator registered. When the mediator from the offer taker is is checked against the local selected mediators it returns null. That is because it tries to check for `localhost:9223` in my case, but it isn't an accepted mediator for Alice. Alice has only the arbitrator `localhost:9222` as mediator selected. So I have the feeling we might run into some issues during this first grace period until updating is required, if mediators are already registered during that time.

chimp1984 commented:
@ripcurlx @mpolavieja 
Did you register the mediator from the same app as the arbitrator? This is required for the update period as otherwise the mediator will be unknown to the old client (old clients had mirrored mediator with arbitrator as it was only prepared code but never implemented - so we need to ensure both have same onion address). Once we have enforced update mediators and arbitrators can (should) be diff. apps (nodes).

I could not reproduce the issue with the correct setup but could reproduce it with diff. arbitrator and mediator. If it was not documented in the setup please update the doc.

ripcurlx commented:
> @ripcurlx @mpolavieja
> Did you register the mediator from the same app as the arbitrator? This is required for the update period as otherwise the mediator will be unknown to the old client (old clients had mirrored mediator with arbitrator as it was only prepared code but never implemented - so we need to ensure both have same onion address). Once we have enforced update mediators and arbitrators can (should) be diff. apps (nodes).
> 
> I could not reproduce the issue with the correct setup but could reproduce it with diff. arbitrator and mediator. If it was not documented in the setup please update the doc.

No, I have used a different data directory for arbitrator and mediator. That explains the problem with the different address. ~So we'll switch to the `real` mediators after the enforced update requirement, correct?~ Didn't read your comment properly ;-)

ripcurlx commented:
https://github.com/bisq-network/bisq/blob/d55114e019fda469266dab50f63916289b2e71cb/core/src/main/java/bisq/core/trade/protocol/tasks/maker/MakerProcessPayDepositRequest.java#L100-L101

Here the error message is wrong. Should be `user.getAcceptedMediatorByAddress(mediatorNodeAddress) must not be null`

I recognized this error when I had the wrong setup with separate arbitrator and mediator

sqrrm reviewed (commented):
ACK

I have tried to break it but only managed to break it by creating a mediation case and then reverting to an older version for one of the traders. Should we handle that case? If we do nothing I think it will be up to the arbitrator to manually create payouts from the look of it, that's still acceptable I think.

chimp1984 commented:
We should state in the release notes very clearly tha downgrading is not supporting and might break existing data structures. Downgrading is basically never supported and tested in Bisq.

chimp1984 commented:
> Here the error message is wrong. Should be user.getAcceptedMediatorByAddress(mediatorNodeAddress) must not be null

Fixed in next bug-fix PR

devinbileck commented:
While testing different variations of upgrade scenarios, I encountered a NullPointer exception. This may not be directly related to mediation (I have yet to confirm), but was encountered while testing it.

**Preconditions**
- Offers of old clients permitted, mediation deactivated:
OfferRestrictions.REQUIRE_UPDATE_DATE = 25.9
MediationManager.MEDIATION_ACTIVATED_DATE = 30.9
- Alice and Bob running old version (v1.1.5).

**Steps to reproduce**
- Bob takes offer, BTC seller as taker.
- Alice confirms payment started.
- Alice updates and starts her client.

**Result**
Alice encounters the following NPE on starting her client.
Notes:
- Did not encounter the error upon restarting the client subsequent times.
- Same error is encountered when Bob updates and starts his client.

```
Sep-10 23:42:31.753 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: Uncaught Exception from thread JavaFX Application Thread
Sep-10 23:42:31.754 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: throwableMessage= null
Sep-10 23:42:31.754 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: throwableClass= class java.lang.NullPointerException
Sep-10 23:42:31.759 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: Stack trace:
java.lang.NullPointerException
	at bisq.core.btc.wallet.WalletService.getTransaction(WalletService.java:610)
	at bisq.core.btc.wallet.WalletService.getTransaction(WalletService.java:615)
	at bisq.core.trade.Trade.getDepositTx(Trade.java:607)
	at bisq.desktop.main.portfolio.pendingtrades.PendingTradesDataModel.doSelectItem(PendingTradesDataModel.java:403)
	at bisq.desktop.main.portfolio.pendingtrades.PendingTradesDataModel.selectBestItem(PendingTradesDataModel.java:379)
	at bisq.desktop.main.portfolio.pendingtrades.PendingTradesDataModel.onListChanged(PendingTradesDataModel.java:374)
	at bisq.desktop.main.portfolio.pendingtrades.PendingTradesDataModel.activate(PendingTradesDataModel.java:151)
	at bisq.desktop.common.model.ActivatableDataModel._activate(ActivatableDataModel.java:28)
	at bisq.desktop.common.model.ActivatableWithDataModel._activate(ActivatableWithDataModel.java:28)
	at bisq.desktop.common.view.ActivatableViewAndModel.lambda$prepareInitialize$0(ActivatableViewAndModel.java:42)
	at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.beans.property.ReadOnlyObjectPropertyBase.fireValueChangedEvent(ReadOnlyObjectPropertyBase.java:74)
	at javafx.beans.property.ReadOnlyObjectWrapper.fireValueChangedEvent(ReadOnlyObjectWrapper.java:102)
	at javafx.scene.Node$ReadOnlyObjectWrapperManualFire.fireSuperValueChangedEvent(Node.java:1054)
	at javafx.scene.Node.invalidatedScenes(Node.java:1114)
	at javafx.scene.Node.setScenes(Node.java:1152)
	at javafx.scene.Parent.scenesChanged(Parent.java:769)
	at javafx.scene.Node.invalidatedScenes(Node.java:1076)
	at javafx.scene.Node.setScenes(Node.java:1152)
	at javafx.scene.Parent$2.onChanged(Parent.java:369)
	at com.sun.javafx.collections.TrackableObservableList.lambda$new$0(TrackableObservableList.java:45)
	at com.sun.javafx.collections.ListListenerHelper$Generic.fireValueChangedEvent(ListListenerHelper.java:329)
	at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
	at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233)
	at javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482)
	at javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541)
	at javafx.collections.ObservableListBase.endChange(ObservableListBase.java:205)
	at javafx.collections.ModifiableObservableListBase.add(ModifiableObservableListBase.java:155)
	at com.sun.javafx.collections.VetoableListDecorator.add(VetoableListDecorator.java:320)
	at com.jfoenix.skins.JFXTabPaneSkin.addTabContentHolder(JFXTabPaneSkin.java:228)
	at com.jfoenix.skins.JFXTabPaneSkin.<init>(JFXTabPaneSkin.java:100)
	at com.jfoenix.controls.JFXTabPane.createDefaultSkin(JFXTabPane.java:67)
	at javafx.scene.control.Control.doProcessCSS(Control.java:897)
	at javafx.scene.control.Control.access$000(Control.java:83)
	at javafx.scene.control.Control$1.doProcessCSS(Control.java:89)
	at com.sun.javafx.scene.control.ControlHelper.processCSSImpl(ControlHelper.java:67)
	at com.sun.javafx.scene.NodeHelper.processCSS(NodeHelper.java:145)
	at javafx.scene.Node.processCSS(Node.java:9529)
	at javafx.scene.Node.processCSS(Node.java:9522)
	at javafx.scene.Node.processCSS(Node.java:9522)
	at javafx.scene.Node.processCSS(Node.java:9522)
	at javafx.scene.Node.processCSS(Node.java:9522)
	at javafx.scene.Scene.doCSSPass(Scene.java:569)
	at javafx.scene.Scene.access$3400(Scene.java:172)
	at javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2477)
	at com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Toolkit.java:412)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:411)
	at com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:438)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:519)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:499)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(QuantumToolkit.java:492)
	at com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(QuantumToolkit.java:320)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

devinbileck commented:
Also while testing the above mentioned upgrade scenarios, I am seeing the following after they update, not sure if it is a concern:
```
Sep-10 23:52:42.719 [JavaFX Application Thread] ERROR bisq.common.proto.ProtoUtil: Invalid value for enum MediationResultState: PB_ERROR_MEDIATION_RESULT
Sep-10 23:52:42.720 [JavaFX Application Thread] ERROR bisq.common.proto.ProtoUtil: We try to lookup for an enum entry with name 'UNDEFINED' and use that if available, otherwise the enum is null. enum=null
```

ripcurlx commented:
> ![Bildschirmfoto 2019-09-11 um 10 05 19](https://user-images.githubusercontent.com/170962/64679420-fd76d480-d47b-11e9-9dba-99cb178c04ec.png)
> @chimp1984 Not critical, but I would remove the `Need help...` in the forth step to prevent unnecessary mediation cases (there shouldn't be any issues at that point that require mediation)

@chimp1984 This is still open.

ripcurlx commented:
> Although it will only be for 10 days as a transitional period that the mediation system will be activated, I think it will cause too much confusion if we have the mediation text in the the "Need help..." and within the popup, but open an arbitration case afterwards.
> ![Bildschirmfoto 2019-09-11 um 10 47 25](https://user-images.githubusercontent.com/170962/64682906-9c063400-d482-11e9-92fa-c21d666ebdf2.png)
> <img alt="Bildschirmfoto 2019-09-11 um 10 53 45" width="1330" src="https://user-images.githubusercontent.com/170962/64682907-9c9eca80-d482-11e9-83c8-6ba2918c17c9.png">

@chimp1984 This is still open.

ripcurlx commented:
<img width="1312" alt="Bildschirmfoto 2019-09-11 um 21 14 39" src="https://user-images.githubusercontent.com/170962/64727466-31ccae00-d4d9-11e9-80fc-39d926f0d20f.png">
I think after a mediation has been closed there should be a message in the final message that the user has to go to the trade again and accept or reject the suggestion made by the mediator. Otherwise people don't know what to do next.

ripcurlx commented:
<img width="1312" alt="Bildschirmfoto 2019-09-11 um 21 16 40" src="https://user-images.githubusercontent.com/170962/64727597-7a846700-d4d9-11e9-9b21-81a75f10da18.png">
After accepting or rejecting a mediation the button still shows `Accept or Reject`. It should be greyed out with `You have accepted` or `You have rejected`

ripcurlx commented:
> Also while testing the above mentioned upgrade scenarios, I am seeing the following after they update, not sure if it is a concern:
> 
> ```
> Sep-10 23:52:42.719 [JavaFX Application Thread] ERROR bisq.common.proto.ProtoUtil: Invalid value for enum MediationResultState: PB_ERROR_MEDIATION_RESULT
> Sep-10 23:52:42.720 [JavaFX Application Thread] ERROR bisq.common.proto.ProtoUtil: We try to lookup for an enum entry with name 'UNDEFINED' and use that if available, otherwise the enum is null. enum=null
> ```

@devinbileck I followed the process described by you (documented in https://bisq.ontestpad.com/script/101#63/2/), but didn't get the exception. 
The only thing I see as well is the error logs mentioned above.

devinbileck commented:
@ripcurlx Now I cannot reproduce the NPE either ;(
